mem/mem.py

```diff
@@ -0,0 +1,80 @@
+from common.configurable_model import ConfigurableModel
+import functools
+from bit_vector import BitVector
+from enum import Enum
+
+
+class Mode(Enum):
+    LINE_BUFFER = 0
+    FIFO = 1
+    SRAM = 2
+
+
+def check_addr(fn):
+    @functools.wraps(fn)
+    def wrapped(self, addr, *args):
+        if not (0 <= addr < self.data_depth):
+            raise ValueError(f"Invalid address ({addr}),"
+                             f" must be within range(0, {self.data_depth})")
+        return fn(self, addr, *args)
+    return wrapped
+
+
+def gen_mem(data_width: int,
+            data_depth: int):
+    class Mem(ConfigurableModel(32, 32)):
+        def __init__(self):
+            super().__init__()
+            self.memory = {}
+            # store self.data_depth so the decorator `check_addr` can reference
+            # it without having to be defined within the scope of the closure
+            self.data_depth = data_depth
+            self.config_addr = BitVector(0, 32)
+
+        def __call__(self):
+            # TODO: Should this define a __call__? What should the semantics
+            # be?
+            raise NotImplementedError()
+
+        @property
+        def mode(self):
```

Comment:
I think `mode` getter should be a private method which we use internally to fetch the mode based on the config register. I think the `mode` setter should not be a public/user-facing function for now.

Reply:
Done